apiserver/objects.go

```diff
@@ -268,3 +258,66 @@ func splitNameAndSHAFromQuery(query url.Values) (string, string, error) {
 	name, sha := charmObjectID[:splitIndex], charmObjectID[splitIndex+1:]
 	return name, sha, nil
 }
+
+// sendJSONError sends a JSON-encoded error response.  Note the
+// difference from the error response sent by the sendError function -
+// the error is encoded in the Error field as a string, not an Error
+// object.
+func sendJSONError(w http.ResponseWriter, req *http.Request, err error) error {
+	if errors.Is(err, errors.NotYetAvailable) {
+		// This error is typically raised when trying to fetch the blob
+		// contents for a charm which is still pending to be downloaded.
+		//
+		// We should log this at debug level to avoid unnecessary noise
+		// in the logs.
+		logger.Debugf("returning error from %s %s: %s", req.Method, req.URL, errors.Details(err))
+	} else {
+		logger.Errorf("returning error from %s %s: %s", req.Method, req.URL, errors.Details(err))
+	}
+
+	perr, status := apiservererrors.ServerErrorAndStatus(err)
+	return errors.Trace(sendStatusAndJSON(w, status, &params.CharmsResponse{
+		Error:     perr.Message,
+		ErrorCode: perr.Code,
+		ErrorInfo: perr.Info,
+	}))
+}
+
+func writeCharmToTempFile(r io.Reader) (string, error) {
+	tempFile, err := os.CreateTemp("", "charm")
+	if err != nil {
+		return "", errors.Annotate(err, "creating temp file")
+	}
+	defer tempFile.Close()
+
+	if _, err := io.Copy(tempFile, r); err != nil {
+		return "", errors.Annotate(err, "processing upload")
+	}
+	return tempFile.Name(), nil
+}
+
+func modelIsImporting(st *state.State) (bool, error) {
+	model, err := st.Model()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return model.MigrationMode() == state.MigrationModeImporting, nil
+}
+
+func emitUnsupportedMethodErr(method string) error {
+	return errors.MethodNotAllowedf("unsupported method: %q", method)
+}
+
+type storageStateShim struct {
+	*state.State
+}
+
+func (s storageStateShim) UpdateUploadedCharm(info state.CharmInfo) (services.UploadedCharm, error) {
+	ch, err := s.State.UpdateUploadedCharm(info)
+	return ch, err
+}
+
+func (s storageStateShim) PrepareCharmUpload(curl string) (services.UploadedCharm, error) {
+	ch, err := s.State.PrepareCharmUpload(curl)
+	return ch, err
+}
```

Comment:
All this was copied over from `apiserver/charm.go`. We now only use these in this file